kdc/bx509d.c

```diff
@@ -1843,6 +1939,189 @@ get_tgt(struct bx509_request_desc *r)
     return ret;
 }
 
+/* Writes one ccache to a response file, as JSON */
+static int
+get_tgts_accumulate_ccache(struct bx509_request_desc *r)
+{
+    const char *fn;
+    size_t bodylen = 0;
+    void *body = NULL;
+    char *body_b64 = NULL;
+    int body_b64len;
+
+    if (r->tgts == NULL) {
+        int fd = -1;
+
+        if (asprintf(&r->tgts_filename,
+                     "%s/tgts-json-XXXXXX", cache_dir) == -1 ||
+            r->tgts_filename == NULL) {
+            free(r->tgts_filename);
+            r->tgts_filename = NULL;
+            return bad_enomem(r, r->error_code = ENOMEM);
+        }
+        if ((fd = mkstemp(r->tgts_filename)) == -1)
+            return bad_req(r, errno, MHD_HTTP_SERVICE_UNAVAILABLE,
+                           "%s", strerror(r->error_code = errno));
+        if ((r->tgts = fdopen(fd, "w+")) == NULL) {
+            (void) close(fd);
+            return bad_req(r, errno, MHD_HTTP_SERVICE_UNAVAILABLE,
+                           "%s", strerror(r->error_code = errno));
+        }
+    }
+    fn = strchr(r->ccname, ':');
+    if (fn == NULL)
+        return bad_req(r, errno, MHD_HTTP_SERVICE_UNAVAILABLE,
+                       "Internal error (invalid credentials cache name)");
+    fn++;
+    if ((r->error_code = rk_undumpdata(fn, &body, &bodylen)))
+        return bad_req(r, errno, MHD_HTTP_SERVICE_UNAVAILABLE,
+                       "%s", strerror(r->error_code));
+    (void) unlink(fn);
+    free(r->ccname);
+    r->ccname = NULL;
+    if (bodylen > INT_MAX >> 4) {
+        free(body);
+        return bad_req(r, errno, MHD_HTTP_SERVICE_UNAVAILABLE,
+                       "Credentials cache too large!");
+    }
+    body_b64len = rk_base64_encode(body, bodylen, &body_b64);
+    if (body_b64len == -1) {
+        free(body);
+        return bad_req(r, errno, MHD_HTTP_SERVICE_UNAVAILABLE,
+                       "%s", strerror(r->error_code = errno));
+    }
+    free(body);
+    /*
+     * XXX Do better -- need to make sure to escape double quotes and control
```

Comment:
Definitely fix this.

Also, write `{"name":"<principal-name>","error":"<error-message>","error_code":<error-code>}` in error cases and _don't_ fail the whole request!

Reply:
Fixed.